hazelcast/src/main/java/com/hazelcast/instance/impl/executejar/ResettableSingleton.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright (c) 2008-2023, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.instance.impl.executejar;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.function.Supplier;
+
+/**
+ * Creates the singleton on the first invocation of {@link #get(Supplier)}  and
+ * afterwards returns the remembered instance.
+ * <p>
+ * The provided {@code Supplier} must return a non-null value.
+ */
+public final class ResettableSingleton<T> {
```

Comment:
why not just ThreadLocal or... simple field? Set once, reset at the end of invocation.

Reply:
ResettableSingleton is a lazy singleton. ThreadLocal or a local field is not appropriate because singleton will be accessed by many threads.  It is reset only by the client at the end of invocation